src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/Runtime/TypeInfos/RuntimeTypeInfo.TypeComponentsCache.cs

```diff
@@ -103,22 +104,38 @@ private static object[] CreatePerNameQueryCaches(RuntimeTypeInfo type, bool igno
             // Generic cache for scenario specific data. For example, it is used to cache Enum names and values.
             internal object? _genericCache;
 
+            // The sole purpose of this struct is to defeat generic code sharing of PerNameQueryCache below.
+            // We eagerly construct PerNameQueryCache when the type components cache is created, but they may not be
+            // actually used in an app that is light on reflection use. But we do have other canonically equivalent
+            // ConcurrentUnifier instances in the program and their use could drag this implementation into the app.
+            // Making the ConcurrentUnifier instance unshareable gives the compiler the ability to unshare this.
+            private struct StringKey : IEquatable<StringKey>
```

Comment:
Apologies for intrusion but could this be a `readonly record struct`? Or is it not allowed in System.Private.CoreLib?

Reply:
We don't use records in CoreLib (or probably all of runtime repo in general). They add _a lot more_ overhead than they're worth. Trading a couple lines of code for kilobytes of footprint for everyone who uses this is not a good tradeoff.

Recently: https://github.com/dotnet/aspnetcore/pull/45604#discussion_r1061091265

Reply:
My assumption was that this could be trimmed but I guess not. Time to read docs on trimming in more detail, thanks for the explanation :)

Reply:
The problem is that a lot of the conveniences records bring is behind interfaces and virtual methods. E.g. `ToString`. It's hard to get rid of `ToString`.